### PR TITLE
Make HomeRobot Great Again!

### PR DESCRIPTION
**Pull request by ferdinandl007**
removed all dependencies of  place note  and integrated native ARKIT  relocalisation and map sharing

### CODE REVIEW

**Comment by cclaan** [comment]
Thanks for putting all this together. I have not been able to test this yet, but I might be able to in the next few days. How good is the map syncing / robot localization with ARKit? Do you have any videos of the AR mode?

**Comment by jhjulkun** [comment]
Thanks! 
I need to dig into code a little bit more, but seems like I'm having issues with the multi peer stuff so if there is a usage video that would be helpful to rule out user error.

I'm getting this though:
peer <MCPeerID: 0x282b38870 DisplayName = iPhone9,4> didChangeState: MCSessionState
HomeRobot[968:132146] [GCKSession] Not in connected state, so giving up for participant [18795B2C] on channel [0].
(and same for channel[1-3]

What combination of iOS&macOS are you guys using ?
I'm using iOS13.2, Catalina 10.15.1 and Xcode 11.2. 
Also: is it better to use the same type of iOS device ?
I'm currently using an iPhone XS and an iPhone7+, I do have two old iPhone 5SE as well, but they are using the A9 cpu vs the A10&A12 which I figured are better for ARKit, but not being the same maybe hurts re-location ?

For iOS13 you need to add this to Info.plist btw, or the app crashes on startup.
`<key>NSBluetoothAlwaysUsageDescription</key>                                                                                     
        <string>Needed for BT in iOS13.</string>`

**Comment by ferdinandl007** [comment]
The map syncing works very well  here is a video of the system working.
https://youtu.be/_FHaIY4Kl9w

**Comment by cclaan** [comment]
Oh that does look pretty good... I tried testing and didn't have as much luck but I probably didn't follow the proper sequence. I'll give it another go later. Apologies on the lousy waypoint navigation code. 
Going to merge this PR. @jhjulkun that multi peer stuff can be finicky, sometimes I have to restart apps a few times.

**Comment by jhjulkun** [comment]
Still having issues, sometimes it works, but most of the time when sending the waypoints the rover goes in the opposite direction. But every now and then it works flawlessly (it seems help if moving the rover to the first waypoint, and once waypoint has been marked green new waypoints usually work).

However, I'm also using my home built rover, so its not unlikely that adds some additional issues..

I'll look at code a little more. It would be nice to clear all waypoints for instance. 

Another question: in the movie the pyramid of the rover itself is close to rover, but for me they often end up floating say 1m up in space ? (and so are waypoints).
Almost seems like the world map/relocation is not working that well ?

Also: would it not be better if rover goes in reverse so that the rear facing camera can see where its going ?

**Comment by cclaan** [comment]
The pyramid should appear on the bot ideally, if not then the map sync between phones isn't perfect. The waypoint stuff is separate from the map sync provided by ARKit. I wrote the waypoint nav in like 20 minutes and it's not very robust. I also may have had my phone facing a different direction from you since I made a custom bracket for it. You might have to negate some values in the code to make it work for you. 
I would like to add another mode where a QR code appears on the screen and you can just scan it to sync coordinate frames, this would avoid having to do any mapping beforehand.

**Comment by jhjulkun** [comment]
Ah: figured out a crucial step when using the app!
After sending map to robot, make sure to move the rover a bit (like in the video) so these show up on the controlling device: 
            addTrackingBall(posPN)

If you never get the tracking balls, it will not work. 

Attaching screenshot of my rover. I'll try cleaning up the code and setup a repository for everything. 

IMG_0026

**Comment by jhjulkun** [comment]
@cclaan Question: 
to create the RMCore and RMShared frameworks, did you create a repository anywhere ? 

Since I originally just thought I could modify just a couple of files I made a joint RMCore/RMShared framework, but I had to more or less pull in every file to make it compile
In short I just build one (big) framework called RMCore that also contains the BLE code to communicate to the rover (its basically sending easy to parse messages using BLE).

In other words, my brute force way to create the framework is probably not the most elegant, but its also a fair amount of work to undo everything, so if you have something ready to go its probably easier to fork that.

**Comment by ferdinandl007** [comment]
https://www.youtube.com/watch?v=v-9B6HPRNEA
ARKit Robot With path planning capabilities
@jhjulkun @cclaan

**Comment by cclaan** [comment]
wow that's great, nice work. Are you using the ARKit feature points or doing some 3d reconstruction?

**Comment by ferdinandl007** [comment]
@cclaan  yes I am using the ARKit feature points :) I got inspired by this [paper](http://openaccess.thecvf.com/content_cvpr_2018_workshops/papers/w30/Xu_Multi-Scale_Voxel_Hashing_CVPR_2018_paper.pdf)  to generate a voxelMap using A hashing function for voxel generation in constant time  this reduces space complexity from n3 to n and allows it to run in real time even in large scale environment.

I also made a coco pod ￼ called [AR￼Navigation￼Kit]( https://github.com/ferdinandl007/ARNavigationKit)￼￼ for this project  it’s pretty easy to use if you want to use it in your own project 😊￼￼￼

**Comment by jhjulkun** [comment]
Very cool! I'm trying to wrap up something else, but I'll try the navigation pod when able.
